tools/heapviz/src/main.rs

```diff
@@ -0,0 +1,90 @@
+use regex::Regex;
+use std::io::{BufRead, Write};
+use std::thread::sleep;
+use std::time::Duration;
+
+struct Event {
+    is_alloc: bool,
+    start: usize,
+    len: usize,
+}
+
+fn main() {
+    let re = Regex::new(r"^(alloc|dealloc)\[(\d+), \d+\] = 0x([0-9a-f]+) \(\d+ ptrs, \d+ bytes\)$")
+        .unwrap();
+
+    let mut events = Vec::new();
+    for line in std::io::stdin().lock().lines() {
+        if let Some(caps) = re.captures(&line.unwrap()) {
+            let typ = caps.get(1).unwrap().as_str();
+            let len = caps.get(2).unwrap().as_str().parse::<usize>().unwrap();
+            let start = usize::from_str_radix(&caps.get(3).unwrap().as_str(), 16).unwrap();
+            events.push(Event {
+                is_alloc: typ == "alloc",
+                start,
+                len,
+            });
+        }
+    }
+
+    let count_alloc = events.iter().filter(|e| e.is_alloc).count();
+    let count_dealloc = events.len() - count_alloc;
+    let start = events.iter().map(|e| e.start).min().unwrap_or(0);
+    let end = events.iter().map(|e| e.start + e.len).max().unwrap_or(0);
+    let mut usage = 0;
+    let peak = events
+        .iter()
+        .map(|e| {
+            if e.is_alloc {
+                usage += e.len;
+            } else {
+                usage -= e.len;
+            }
+            usage
+        })
+        .max()
+        .unwrap_or(0);
+    let len = end - start;
+    println!(
+        "Observed {} allocations and {} deallocations",
+        count_alloc, count_dealloc
+    );
+    println!("Start address: {:08x}", start);
+    println!("End address: {:08x}", end);
+    println!("Peak usage: {0} = {0:08x} bytes", peak);
+    println!("Peak consumption: {0} = {0:08x} bytes", len);
+    println!("Fragmentation overhead: {0} = {0:08x} bytes", len - peak);
+
+    print!("\nLoading visualization");
+    for _ in 0..30 {
```

Comment:
It's meant to leave some time to see the previous output (which will also be visible afterwards when the visualization is over, or when interrupting it with CTRL+C). Ideally there should be some "Press any key to continue", although `stdin` is already captured to parse the input. I can check if `ncurses` could help capturing the ENTER key here. Otherwise I think it's fine to leave it like this, or to start the visualization immediately - this tool isn't meant to be ultra polished at this point.

Reply:
After taking a brief look, `ncurses` reads input from `stdin` as well. So the tool could read from a file instead of stdin, but that would make it more complex than it is needed for the purpose (a simple tool to help debug heap usage).

Reply:
Probably worth adding a short comment to justify this loop as at first glance it seems useless :)

Reply:
I refactored the code with some small CLI after all. It's now reading the log from a file and waiting for the user to press ENTER to start the visualization.